### ISSUE TITLE
setTerrain is missing

### ISSUE THREAD

**Issue by openSourcerer9000** [mapcomponents/react-map-components-maplibre#14]
It seems [MapLibreGlWrapper ](https://github.com/mapcomponents/react-map-components-maplibre/blob/0ebf8a566a3a5cee326eccecb3c8fbd979ff0875/src/components/MapLibreMap/lib/MapLibreGlWrapper.js) does not pass the full `map` object API from maplibre. How else can I access the map? useMap doesn't seem to work for me at all. I'm not sure which method is preferred.

@mapcomponents/react-maplibre version: 
0.1.27

## Steps To Reproduce

```
//terrain3d.js
import {useContext,useEffect} from 'react'
import { MapContext } from "@mapcomponents/react-core"
import {
    useMap
  } from "@mapcomponents/react-maplibre";
  
const Terrain3D = (props) => {
    const mapContext = useContext(MapContext);
    // const mapHook = useMap({ mapId: "map_1" });
    const {lyrDir=process.env.PUBLIC_URL,vis=true,id='Terrain3D',lyrSuffix='png',
        mapId="map_1",
        insertBeforeLayer=mapContext.map?.firstSymbolLayer,
        minzoom=4,maxzoom=18,
    } = props

    useEffect(() => {
        if (mapContext.map?.isStyleLoaded) {
            console.log(mapContext.map)
            mapContext.map?.addSource("terrain-dem", {
                type: "raster-dem",
                tiles: [
                    `${lyrDir}/${id}/{z}/{x}/{y}.${lyrSuffix}`
                ],
                tileSize: 256,
                maxzoom: 18,
                encoding: "terrarium",
            });
            mapContext.map?.setTerrain(
                { source: "terrain-dem", exaggeration: 20 });

        }
    }, [mapContext.map?.isStyleLoaded])
    
    

  return (
    <>

    </>
  )
}

export default Terrain3D

```

## The current behavior
Just add this component to a map. Printing to the console shows there is indeed no setTerrain method available, although it is available from vanilla maplibre gl. No need to create your own terrain tiles, trying to call setTerrain results in 

```
terrain3D.js:40 Uncaught TypeError: _mapContext$map4.setTerrain is not a function
    at Terrain3D._mapContext$map5 (terrain3D.js:40:1)
    at invokePassiveEffectCreate (react-dom.development.js:23487:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
    at invokeGuardedCallback (react-dom.development.js:4056:1)
    at flushPassiveEffectsImpl (react-dom.development.js:23574:1)
    at unstable_runWithPriority (scheduler.development.js:468:1)
    at runWithPriority$1 (react-dom.development.js:11276:1)
    at flushPassiveEffects (react-dom.development.js:23447:1)
    at react-dom.development.js:23324:1
    at workLoop (scheduler.development.js:417:1)
    at flushWork (scheduler.development.js:390:1)
    at MessagePort.performWorkUntilDeadline (scheduler.development.js:157:1)
Terrain3D._mapContext$map5 @ terrain3D.js:40
invokePassiveEffectCreate @ react-dom.development.js:23487
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
flushPassiveEffectsImpl @ react-dom.development.js:23574
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushPassiveEffects @ react-dom.development.js:23447
(anonymous) @ react-dom.development.js:23324
workLoop @ scheduler.development.js:417
flushWork @ scheduler.development.js:390
performWorkUntilDeadline @ scheduler.development.js:157
index.js:1 The above error occurred in the <Terrain3D> component:

    at Terrain3D (http://localhost:3000/static/js/main.chunk.js:2070:76)
    at MapComponentsProvider (http://localhost:3000/static/js/vendors~main.chunk.js:7001:23)
    at Mapp (http://localhost:3000/static/js/main.chunk.js:894:76)
    at InnerThemeProvider (http://localhost:3000/static/js/vendors~main.chunk.js:117497:72)
    at ThemeProvider (http://localhost:3000/static/js/vendors~main.chunk.js:115773:24)
    at ThemeProvider (http://localhost:3000/static/js/vendors~main.chunk.js:117516:24)
    at StyledEngineProvider (http://localhost:3000/static/js/vendors~main.chunk.js:116066:27)
    at App
    at MapComponentsProvider (http://localhost:3000/static/js/vendors~main.chunk.js:7001:23)
    at AppContextProvider (http://localhost:3000/static/js/main.chunk.js:280:23)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries
```

## The expected behavior
Adding terrain to the referenced map at `mapId`.

**Comment by openSourcerer9000** [comment]
Could `setTerrain` simply be added to this list?

https://github.com/mapcomponents/react-map-components-maplibre/blob/0ebf8a566a3a5cee326eccecb3c8fbd979ff0875/src/components/MapLibreMap/lib/MapLibreGlWrapper.js#L437-L476

**Comment by cioddi** [comment]
It is best to use `mapHook.map.map` or `mapContext.getMap(mapId).map` to access the maplibre instance directly.
I am not sure whether we are going to keep the duplicated functions accessible on the wrapper instance. The API is likely to change in that regard but the maplibre instance of course is safe.

The setTerrain function (a feature I am really looking forward to) has been merged 6 days ago into the maplibre main branch but is not included in a stable release yet. It is included in a prerelease you can add to your project using `yarn add maplibre-gl@v2.2.0-pre.2`. After running that you should be able to access setTerrain on both mapHook.map and mapHook.map.map.

**Comment by openSourcerer9000** [comment]
Yes, it looks like the terrain should show up in `mapHook.map.map.style.terrain` but it is absent after calling setTerrain.
https://docs.mapbox.com/mapbox-gl-js/style-spec/terrain/#source

**Comment by cioddi** [comment]
Hi @openSourcerer9000,
I reccomend using the useMap hook. The "map" Protperty it return ist a state variable that will become truthy once the maplibre instance is ready for interaction.

I have created a demo locally that works well with the following Terrain Component:


```
import { useEffect } from "react";
import { useMap } from "@mapcomponents/react-maplibre";

const TerrainLayer = (props) => {
  const mapHook = useMap({ mapId: "map_1" });

  useEffect(() => {
    if (!mapHook.map) return;

    if (!mapHook.map.map.getSource("terrain")) {
      mapHook.map.map.addSource("terrain", {
        type: "raster-dem",
        tiles: ["https://vtc-cdn.maptoolkit.net/terrainrgb/{z}/{x}/{y}.webp"],
        encoding: "mapbox",
        maxzoom: 14,
        minzoom: 4,
      });
    }
    mapHook.map.map.setTerrain({ source: "terrain", exaggeration: 1 });
  }, [mapHook.map]);

  return <></>;
};

export default TerrainLayer;
```

I am currenty very busy but I will create a working demo repository later today so you can test it locally.

**Comment by openSourcerer9000** [comment]
Hm, I'm not sure why that's working on your end. That's what I tried first, this is it's complaint: `Uncaught TypeError: mapHook.map.map.setTerrain is not a function`. On my end, only the mapHook.map `<MapLibreGLWrapper/>` object exposes setTerrain. This is my mapHook.map.map logged to the console:

![image](https://user-images.githubusercontent.com/61931826/169302039-7477c983-6450-4226-a156-1f356f427623.png)

**Comment by cioddi** [comment]
I have created a demo repository here https://github.com/cioddi/mapcomponents-setterrain-demo

Please check if that works for you. I have included a copy of the MapLibreMap component to make sure it picks up the. A possibility to choose the maplibre-gl version you want should be added.

To make it look nice (as seen in the screenshots of the PR on the maplibre-gl repository) we need to add a hillshade layer, which I haven't done yet in the demo repo.

**Comment by openSourcerer9000** [comment]
OK thanks, yes bringing that MapLibreMap in locally worked for me. Unfortunately, it's removing the closer raster tiles from view when terrain is on, the higher the exaggeration, the worse the problem. Something to do with the camera being distorted from the viewport? It's like it thinks the closest tiles are out of view. 
![image](https://user-images.githubusercontent.com/61931826/169377221-8f115436-3769-401b-9835-922f9fef37d0.png)

I can recreate this in your example repo as well to some extent. Replace the vector basemap with satellite imagery, I'm using `https://api.maptiler.com/maps/hybrid/style.json?key={maptilerAPIkey}`, and bump up the exaggeration to 15. Mountain locations do not have this problem, just flat areas. Panning over to South Louisiana/Texas, US, you can see the same phenomenon:
![image](https://user-images.githubusercontent.com/61931826/169380792-13628eaa-ba1d-411f-a3bf-04f8973d7a8b.png)

**Comment by cioddi** [comment]
I don't know about the valid value range for the exaggeration option. All the code contained in the PR on maplibre-gl suggest values between 0 and 1. Higher values might not trigger error in the console but as seen in the visual output. Since this feature is not yet officially documented it might be just that high value. Similar to the, now possible, higher pitch setting, it may lead to unexpected results.

If you upload a repository with you application code I can take a look at it, maybe I can spot another reason for this behavior.

**Comment by openSourcerer9000** [comment]
OK, see below. I described what I was seeing in the README

https://github.com/openSourcerer9000/mapcomponents-setterrain-demo

**Comment by cioddi** [comment]
Your Repository works fine for me, for the lower zoom levels. 
I start seeing something similiar to what is shown in the screenshots at zoom 16.54, the meshes too close to the camera aren't rendered leaving the spot black. I was about to suggest opening an issue in the maplibre-gl repository, but I see you have already done that https://github.com/maplibre/maplibre-gl-js/issues/1241 

If you set the exaggeration to .6 and zoom out a bit this is a beautiful and fun demo, with some performance issues that will hopefully be resolved in maplibre-gl. So far everyboy involved seems to have made an amazing job. 

To use a pitch value higher than 60 you also need to set maxPitch to the desired value. I do not recommend setting it higher than 65 as it dramatically increases the amount of tiles the have to be rendered.

**Comment by openSourcerer9000** [comment]
Yeah it's just a bummer it doesn't render properly for flat, low lying areas. For our purposes we need to inspect small changes in a DEM, and with the flatness of the area it requires a Z-factor of at least 10. So it just comes out like the screenshot above where less than a third of the viewport renders

**Comment by cioddi** [comment]
Actually, now that you pointed that out I took a closer look at that screenshot. It reminds me of something I have often seen in blender, esp. When using blenderGIS. I think it is an issue with the web gl perspective settings.

If you take a look at the first example on this website https://webglfundamentals.org/webgl/lessons/webgl-3d-perspective.html

Reduce the fudge factor to 0.0 and then reduce the z value to about -155. You get the same effect where meshes are cleanly cut off. This is different from the other issue, where meshes are rendered with incomplete textures but eventually will be rendered correctly as seen in the video you posted on the maplibre issue.
